### PR TITLE
Add msi installer

### PR DESCRIPTION
**Pull request by joaquimrocha**
## Testing
Install Wix and ensure that light.exe and candle.exe are in the Path.
- [ ] Run: `make`, then move to the app folder and run `npm run package-msi`: Test install the MSI package that's been generated in the dist/ folder.

### CODE REVIEW

**Comment by joaquimrocha** [comment]
We should add a Github action later, to generate the MSI as well as the regular NSIS.

File: app/windows/msi/build.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+});
+
+const msiOptions = {
+  appDirectory: APP_DIR,
+  outputDirectory: OUT_DIR,
+  description: info.description,
+  exe: exeName,
+  name: nameOptions.productName,
+  manufacturer: info.author.name,
+  version: process.env.npm_package_version,
```
Comment (illume) [inline app/windows/msi/build.js:30]: I'm not sure about `process.env.npm_package_version`. Can you please explain why this is used?

File: app/windows/msi/build.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  arch: 'x86',
+};
+
+// Generate the exe name from electron-builder's artifactName
+let exeName = info.build.artifactName.split('.')[0];
+Object.entries(nameOptions).forEach(([key, value]) => {
+  exeName = exeName.replace(`\${${key}}`, value);
+});
+
+const msiOptions = {
```
Comment (illume) [inline app/windows/msi/build.js:23]: I reviewed the configuration options which are documented here: https://github.com/felixrieseberg/electron-wix-msi#configuration

> shortName (optional, string) - A short name for the app, used wherever spaces and special characters are not allowed. Will use the name if left undefined.

I wonder if we need a shortName here? It's a place for where spaces and special characters are not allowed. For "Headlamp" it is ok to not have one, but for names like "Sxxxx Bxxxx" it's not.

> upgradeCode (string, optional) - A unique UUID used by your app to identify itself. This module will generate one for you, but it is important to reuse it to enable conflict-free upgrades.

This `upgradeCode` seems important to keep track of and check.

> arch (string, optional) - Defines the architecture the MSI is build for. Values can be either x86 or x64. Default's to x86 if left undefined.

How is arch handled here? We need a x64 only? Seems we are setting 'x86'.

Perhaps we could override some of the images shown in the installer?https://github.com/felixrieseberg/electron-wix-msi#ui-configuration-optional

File: app/windows/msi/build.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
+// You need to install wix tool set first
+// Reference https://ourcodeworld.com/articles/read/927/how-to-create-a-msi-installer-in-windows-for-an-electron-framework-application
```
Comment (illume) [inline app/windows/msi/build.js:2]: Can you please add the docs link?
https://github.com/felixrieseberg/electron-wix-msi#configuration

**Review by illume — CHANGES_REQUESTED** [review]
Looking good so far. I left some questions and notes.

For the first commit, can you please add the `app:` prefix?

File: app/windows/msi/build.js (hunk 1 above)
Comment (joaquimrocha) [reply app/windows/msi/build.js:30]: This is just the app's version. We can use the info.version instead, though it should be the same.

File: app/windows/msi/build.js (hunk 2 above)
Comment (joaquimrocha) [reply app/windows/msi/build.js:23]: All good points.
I will address all but the images, which we can add later.

**Comment by joaquimrocha** [comment]
Hopefully it's good now.

File: docs/development/_index.md [hunk 4]
```diff
@@ -66,11 +66,18 @@ make app-linux
...
+make app-mac
 ```
 
 ```bash
-make app-mac
+make app-win
+```
+
+For Windows, by default it will produce an installer using [NSIS (Nullsoft Scriptable Install System)](https://sourceforge.net/projects/nsis/).
+If you prefer an `.msi` installer, then be sure to install the [WiX toolset](https://wixtoolset.org/), have its `light.exe` and `candle.exe` in the Windows path, and run:
```
Comment (illume) [inline docs/development/_index.md:77]: I have wixtoolset installed with chocolatey, but there is no light.exe or candle.exe. Do you know where it should be?

Comment (illume) [reply docs/development/_index.md:77]: I guess the chocolatey install was broken. The installer worked.

Comment (joaquimrocha) [reply docs/development/_index.md:77]: I installed with WiX but had to add those executables to Windows's Path.

File: docs/development/_index.md [hunk 5]
```diff
@@ -66,11 +66,18 @@ make app-linux
...
+
+```bash
```
Comment (illume) [inline docs/development/_index.md:79]:
```
set PATH=%PATH%;C:\Program Files (x86)\WiX Toolset v3.11\bin
make app-win-msi
```

**Comment by illume** [comment]
There's no branding/headlamp logo on the MSI installer:
(but I think you know that?)
image

Here's a screenie of the uninstaller without the logo:
image

Note: no icon on msi file in explorer:
image

The installer doesn't add a startbar shortcut, but does add a desktop icon.

---

The installer works, but I cannot run the Headlamp program (from desktop shortcut or from cmd line).  There's no error message (or any logging at all).

Does it work for you?

**Comment by joaquimrocha** [comment]
> The installer works, but I cannot run the Headlamp program (from desktop shortcut or from cmd line). There's no error message (or any logging at all).
> 
> Does it work for you?

It did work from me. I launched it from the entry in the Start menu *I think*. I will try again.

As for the lack of branding. This is not an installer we're publishing ATM yet, so we can add all those images later.

**Comment by joaquimrocha** [comment]
@illume , I have verified that there was indeed an issue with running the app after installation. It had to do with the `exe` name given, which I thought was just the installer name but it's the app's executable. I updated that and left a comment.

I also don't see a start menu now. Not sure why... But this is a detail I think we can refine later.

**Comment by illume** [comment]
> I also don't see a start menu now. Not sure why... But this is a detail I think we can refine later.

Cool. What's left open for the MSI based installer?
- protocol support (adding the registry key you mentioned)
- branding
- start menu item (I believe it had a start menu item for me... so I don't know what's going on there)

Should these be done in this PR or other ones?

File: docs/development/_index.md (hunk 5 above)
Comment (illume) [reply docs/development/_index.md:79]: Would you mind adding this set PATH command in to the docs there?

An improvement for later might be to check if WiX Toolset is there and add it to the path inside the make target/or npm script.

**Comment by joaquimrocha** [comment]
> > I also don't see a start menu now. Not sure why... But this is a detail I think we can refine later.
> 
> Cool. What's left open for the MSI based installer?
> 
> * protocol support (adding the registry key you mentioned)
> * branding
> * start menu item (I believe it had a start menu item for me... so I don't know what's going on there)

Yes.

> Should these be done in this PR or other ones?

I think this PR can already be useful for some people, and has minimal impact to the UX since we're not publishing the .msi just now. So let's merge it in and I will open an issue to track the remaining tasks, after we merge this PR.

**Pull request merged**